_posts/2017-09-08-scalafix-v0.5.md

```diff
@@ -0,0 +1,190 @@
+---
+category: blog
+by: Ólafur Páll Geirsson
+title: "Announcing Scalafix v0.5"
+---
+
+I am excited to announce the release of [Scalafix v0.5.0][].
+This release introduces new features in addition to several bug fixes and
+integration improvements.
+
+If you are not familiar with Scalafix, Scalafix is a rewrite and linting tool
+for Scala.
+The project is developed at the [Scala Center][sc] with the mission to help
+automate migration between different Scala compiler and library versions.
+
+> This is the third post on Scalafix and scalameta.
+> You might be interested in reading the previous post
+> [here](http://www.scala-lang.org/blog/2017/02/27/Scalafix-v0.3.html).
+
+
+## Linting
+
+Scalafix v0.5 supports linters.
+Before v0.5, Scalafix rules could only "fix" code by providing rewrite instructions.
+Now, it's possible to implement Scalafix rules that report lint messages instead
+of rewrite instructions.
+Each category of lint messages is attached to a default severity level
+(info/warn/error), which users can override via configuration.
+
+The first Scalafix linter is [`NoInfer`][].
+`NoInfer` reports errors when the Scala compiler infers one of the following types:
+`Any`, `AnyVal`, `Product` or `Serializable`.
+It's usually a bad sign when the compiler infers such a generic type.
+In those cases, it's better to explicitly type annotate the expected type.
+
+
+### Building a custom linter
+It's simple to implement custom Scalafix linters.
+To demonstrate this, let's implement a `NoDefaultParams` linter that reports an error if
+default parameter values are used.
+To begin with, we motivate this linter rule with a bit of documentation.
+
+<script src="https://gist.github.com/olafurpg/38699f591329c3bcff52b7412dcb696b.js"></script>
+
+Looks good, now we can implement the Scalafix rule.
+Rules are implemented with Scalafix and Scalameta.
+The `param"..."` part is a [quasiquote][], which is a simple way to construct
+or pattern-match a Scalameta tree from syntax.
+
+<script src="https://gist.github.com/olafurpg/7b022c7120156716def282028b091e8b.js"></script>
+
+When `NoDefaultParams` encounters a parameter with a default parameter value,
+it will report a message like this
+
+```scala
+scala/test/NoDefaultParams.scala:8: error: [NoDefaultParams] default parameter values are not allowed.
+  def foo(e: Int = 2) = e
+                   ^
+```
+
+We can use Scalafix-testkit to test that our rewrite works as expected.
```

Comment:
What do you think about advertising the existence of the g8 template project? It takes care of setting up the test harness, and I found it incredibly useful apart from a couple minor things.

Reply:
Good idea, done.